fe/fe-core/src/main/java/com/starrocks/sql/optimizer/SubqueryUtils.java

```diff
@@ -172,4 +123,31 @@ private static boolean checkPredicateContainColumnRef(List<ColumnRefOperator> cr
 
         return false;
     }
+
+    /**
+     * rewrite the predicate and collect info according to your operatorShuttle
+     * @param correlationPredicate
+     * @param scalarOperatorShuttle
+     * @return
+     */
+    public static ScalarOperator rewritePredicateAndExtractColumnRefs(
+            ScalarOperator correlationPredicate, BaseScalarOperatorShuttle scalarOperatorShuttle) {
+        if (correlationPredicate == null) {
+            return null;
+        }
+        return correlationPredicate.clone().accept(scalarOperatorShuttle, null);
+    }
+
+    public static boolean containsExpr(Map<ColumnRefOperator, ScalarOperator> innerRefMap) {
+        return innerRefMap.values().stream().anyMatch(e -> !e.isColumnRef());
+    }
```

Comment:
emmm, not recommended extract to a method...inline may be easier to understand.

Reply:
It's a common action in all the apply to join rules. I think it's a bit of tedious to write it four places. So if inline is our style, I will change it back.

Reply:
Inline isn't our style, the name of this method isn't clear, isn't as good as inline

Reply:
yeh. I renamed the method to express its usage more explicit.